packages/types/src/config/config.ts

```diff
@@ -0,0 +1,181 @@
+    rococo = 'rococo',
+    kusama = 'kusama',
+    polkadot = 'polkadot',
+    shiden = 'shiden',
+}
+
+const EnvironmentTypesSchema = z.nativeEnum(EnvironmentTypes)
+
+export enum NetworkNames {
+    development = 'development',
+    rococo = 'rococo',
+    kusama = 'kusama',
+    polkadot = 'polkadot',
+    shiden = 'shiden',
+}
+
+const NetworkNamesSchema = z.nativeEnum(NetworkNames)
+
+export const DatabaseConfigSchema = z
+    .record(
+        EnvironmentTypesSchema,
+        z.object({
+            type: z.nativeEnum(DatabaseTypes),
+            endpoint: z.string(),
+            dbname: z.string(),
+            authSource: z.string(),
+        })
+    )
+    .optional()
+
+export const BatchCommitConfigSchema = z.object({
+    interval: z.number().positive(),
+    maxBatchExtrinsicPercentage: z.number().positive(),
+})
+
+export type BatchCommitConfig = z.infer<typeof BatchCommitConfigSchema>
+
+export type DatabaseConfig = z.infer<typeof DatabaseConfigSchema>
+
+export const ProsopoBaseConfigSchema = z.object({
+    logLevel: z.nativeEnum(LogLevel),
+    defaultEnvironment: z.nativeEnum(EnvironmentTypes).default(EnvironmentTypes.development),
+    // The account with which to query the contract and sign transactions
+    account: z.object({
+        address: z.string(),
+        secret: z.string().optional(),
+        password: z.string().optional(),
+    }),
+})
+
+export const NetworkConfigSchema = z.object({
+    endpoint: z.string().url(),
+    contract: z.object({
+        address: z.string(),
+        name: z.string(),
+    }),
+    accounts: z.array(z.string()),
+})
+
+export type NetworkConfig = z.infer<typeof NetworkConfigSchema>
+
+export const ProsopoNetworksSchema = z.record(NetworkNamesSchema, NetworkConfigSchema.required()).default({
+    development: {
+        endpoint: 'ws://127.0.0.1:9944',
+        contract: {
+            address: '',
+            name: '',
+        },
+        accounts: [''],
+    },
+})
+
+export const ProsopoBasicConfigSchema = ProsopoBaseConfigSchema.merge(
+    z.object({
+        networks: ProsopoNetworksSchema,
+        database: DatabaseConfigSchema.optional(),
+    })
+)
+
+export type ProsopoBasicConfig = z.infer<typeof ProsopoBasicConfigSchema>
+
+export const ProsopoCaptchaCountConfigSchema = z.object({
+    solved: z.object({
+        count: z.number().positive(),
+    }),
+    unsolved: z.object({
+        count: z.number().nonnegative(),
+    }),
+})
+
+export const ProsopoImageServerConfigSchema = z.object({
+    baseURL: z.string().url(),
+    port: z.number().optional().default(9229),
+    fileServePaths: z.string(),
+})
+
+export const ProsopoCaptchaSolutionConfigSchema = z.object({
+    requiredNumberOfSolutions: z.number().positive().min(2),
+    solutionWinningPercentage: z.number().positive().max(100),
+    captchaFilePath: z.string(),
+    captchaBlockRecency: z.number().positive().min(2),
+})
+
+export const ProsopoServerConfigSchema = ProsopoBasicConfigSchema.merge(
```

Comment:
can we make `ProsopoServerConfigSchema` and `ProsopoClientConfigSchema` inherit from a shared config at all? It can be private. There's a lot of shared fields between the two of them so it makes sense to have a single declaration, especially considering they need to be the same as each other (e.g. one shouldn't have `userAccountAddress` while the other has `userAccount` both using those fields for the same thing, we should force them to be the same name for consistency, hence shared fields)

This is why I used the omit function to remove a couple of fields from the other type, however I prefer these types with a shared set of fields

Reply:
Done